twos-complement/bigint2twos.go

```diff
@@ -0,0 +1,82 @@
+package twoscomplement
+
+import "math/big"
+
+var bigOne = big.NewInt(1)
+
+// ToBytes returns a variable length two's complement byte array representation of the input.
+// Does not alter input.
+func ToBytes(bi *big.Int) []byte {
+	var resultBytes []byte
+	switch bi.Sign() {
+	case -1:
+		// compute 2's complement
+		plus1 := big.NewInt(0)
+		plus1 = plus1.Add(bi, big.NewInt(1)) // add 1
+		resultBytes = plus1.Bytes()
+		for i, b := range resultBytes {
+			resultBytes[i] = ^b // negate every bit
+		}
+		if len(resultBytes) == 0 || resultBytes[0]>>7 != 1 {
```

Comment:
If `len(resultBytes) == 0`, shouldn't we return `[]byte{}`? Can this even be true here, i.e. if `bi.Sign()` is nonzero?

Reply:
Yes, for -1. -1 + 1 = 0.

Reply:
Probably this is why wikipedia recommends adding 1 _after_ inverting the bits, to avoid these edge cases.